helpers/misc.go

```diff
@@ -140,31 +136,42 @@ func IsNotZeroAndNotEqual[T any](given T, equal T) bool {
 		return true
 	}
 
-	for i := 0; i < givenValue.NumField(); i++ {
+	for i := range givenValue.NumField() {
 		if !givenValue.Field(i).IsZero() &&
 			givenValue.Field(i).CanInterface() &&
 			givenValue.Field(i).Interface() != equalValue.Field(i).Interface() {
-
 			return true
 		}
 	}
 	return false
 }
 
 // MergeNonZero is used to merge non-zero overrides from one struct into another of the same type
-func MergeNonZero[T any](original T, overrides T) T {
-	originalValue := reflect.ValueOf(&original)
-	overridesValue := reflect.ValueOf(&overrides)
+func MergeNonZero[T any](original, overrides T) T {
+	// Create a copy of original that we'll modify
+	result := original
 
-	for i := 0; i < originalValue.Elem().NumField(); i++ {
-		if !overridesValue.Elem().Field(i).IsZero() &&
-			overridesValue.Elem().Field(i).CanSet() {
+	// Get reflect values, using the actual values not pointers to them
+	resultValue := reflect.ValueOf(&result).Elem()
+	overridesValue := reflect.ValueOf(overrides)
 
-			overrideField := overridesValue.Elem().Field(i)
-			originalValue.Elem().Field(i).Set(overrideField)
+	// Ensure we're working with structs
+	if resultValue.Kind() != reflect.Struct || overridesValue.Kind() != reflect.Struct {
+		return original // Can't merge non-structs
+	}
+
+	// Iterate through fields
+	for i := range resultValue.NumField() {
+		resultField := resultValue.Field(i)
+		overrideField := overridesValue.Field(i)
+
+		// Check if override field is non-zero and result field can be set
+		if !overrideField.IsZero() && resultField.CanSet() {
+			resultField.Set(overrideField)
 		}
 	}
-	return originalValue.Elem().Interface().(T)
+
+	return result
```

Comment:
This behaviorally changes this function and is a breaking change to only operate on structs, but I believe that was the original intent of this function and will not break any current functionality.